autotest/ogr/ogr_xodr.py

```diff
@@ -0,0 +1,285 @@
+
+
+def ogr_xodr_check_road_signal_geometry_dissolve(lyr, dissolve_tin: bool):
+    lyr.ResetReading()
+    feat = lyr.GetNextFeature()
+    wkt = feat.GetGeometryRef().ExportToWkt()
+    if not dissolve_tin:
+        assert (
+            wkt
+            == "TIN Z (((618366.844654328 5809540.96164437 103.568946384872,618366.840967264 5809541.48457345 103.54861591048,618367.044614501 5809540.96290705 103.56516023851,618366.844654328 5809540.96164437 103.568946384872)),((618366.840967264 5809541.48457345 103.54861591048,618367.040927437 5809541.48583613 103.544829764117,618367.044614501 5809540.96290705 103.56516023851,618366.840967264 5809541.48457345 103.54861591048)),((618366.858657359 5809540.99087441 104.318245603892,618367.058617531 5809540.99213709 104.31445945753,618366.854970294 5809541.51380349 104.297915129499,618366.858657359 5809540.99087441 104.318245603892)),((618366.854970294 5809541.51380349 104.297915129499,618367.058617531 5809540.99213709 104.31445945753,618367.054930467 5809541.51506617 104.294128983137,618366.854970294 5809541.51380349 104.297915129499)),((618366.854970294 5809541.51380349 104.297915129499,618367.054930467 5809541.51506617 104.294128983137,618366.840967264 5809541.48457345 103.54861591048,618366.854970294 5809541.51380349 104.297915129499)),((618366.840967264 5809541.48457345 103.54861591048,618367.054930467 5809541.51506617 104.294128983137,618367.040927437 5809541.48583613 103.544829764117,618366.840967264 5809541.48457345 103.54861591048)),((618367.058617531 5809540.99213709 104.31445945753,618366.858657359 5809540.99087441 104.318245603892,618367.044614501 5809540.96290705 103.56516023851,618367.058617531 5809540.99213709 104.31445945753)),((618367.044614501 5809540.96290705 103.56516023851,618366.858657359 5809540.99087441 104.318245603892,618366.844654328 5809540.96164437 103.568946384872,618367.044614501 5809540.96290705 103.56516023851)),((618366.844654328 5809540.96164437 103.568946384872,618366.858657359 5809540.99087441 104.318245603892,618366.854970294 5809541.51380349 104.297915129499,618366.844654328 5809540.96164437 103.568946384872)),((618366.854970294 5809541.51380349 104.297915129499,618366.840967264 5809541.48457345 103.54861591048,618366.844654328 5809540.96164437 103.568946384872,618366.854970294 5809541.51380349 104.297915129499)),((618367.044614501 5809540.96290705 103.56516023851,618367.054930467 5809541.51506617 104.294128983137,618367.058617531 5809540.99213709 104.31445945753,618367.044614501 5809540.96290705 103.56516023851)),((618367.044614501 5809540.96290705 103.56516023851,618367.040927437 5809541.48583613 103.544829764117,618367.054930467 5809541.51506617 104.294128983137,618367.044614501 5809540.96290705 103.56516023851)))"
+        ), "wrong geometry created for RoadSignal"
+    else:
+        assert (
+            wkt == "POINT (618366.942790883 5809541.22374025 103.556888074495)"
+        ), "wrong geometry created for dissolved RoadSignal"
```

Comment:
It would be worth adding a test that runs the "test_ogrsf" compliance checking utility on the sample file. Something like:
```
def test_ogr_xodr_test_ogrsf():

    import test_cli_utilities

    if test_cli_utilities.get_test_ogrsf_path() is None:
        pytest.skip()

    ret = gdaltest.runexternal(
        test_cli_utilities.get_test_ogrsf_path() + " -ro " + xodr_file
    )

    assert "INFO" in ret
    assert "ERROR" not in ret
```

Reply:
Oh, that's a good one 😅! After adding this test, a whole lot of new errors are exposed.

Reply:
Haven't consciously worked with OGR spatial and attribute filters yet. Apache Arrow streams also not yet considered. I will dig into that.

```
INFO: Testing layer ReferenceLine.
...
E           ERROR: Attribute filter seems to be ignored in GetFeatureCount() when spatial filter is set.
E         ? +++++
E           ERROR: Spatial filter (0) failed to eliminatea feature unexpectedly!
...
E           ERROR: GetFeatureCount() may not be taking attribute filter into account (nInclusiveCount = 41, nExclusiveCount = 41, nExclusiveCountWhileIterating = 0, nTotalCount = 41).
...
E           ERROR: GetArrowStream() with 1=0 filter returned 41 features, whereas 0 expected
...
E           ERROR: GetArrowStream() with ID <> '2233498' filter returned 41 features, whereas less than 41 expected
...
E           ERROR: ExecuteSQL() should have returned a layer without features.
E           ERROR: ExecuteSQL() should have returned a layer without features.

```

Reply:
As we are not explicitly addressing spatial and attribute filters in any of our layers, I suspect that we are still not implementing our layers 100 % correctly. Otherwise these tests which fall back to the default implementation of `OGRLayer` should by default run correctly, I suppose.

Reply:
> Apache Arrow streams also not yet considered.

don't worry about those ones. They are just a consequence of regular feature-based filtering not working yet, as your driver relies on the generic GetArrowArray implementation that fallbacks to using GetNextFeature() and is thus subject to its bugs

> As we are not explicitly addressing spatial and attribute filters in any of our layers,

Instead of directly implementin GetNextFeature(), you could could rename it to GetNextRawFeature() and make your OGRxxxxLayer classes extend from ``public OGRLayer, public OGRGetNextFeatureThroughRaw<OGRxxxxLayer>``. Cf for example at ogr/ogrsf_frmts/lvbag/ogr_lvbag.h . The OGRGetNextFeatureThroughRaw class implements a GetNextFeature() method that calls GetNextRawFeature() and then evaluates the spatial and attribute filters

Reply:
I appreciate your precise guidance. Following your suggestion, everything works as expected! The new test is passing and it saved me probably ~~half~~ a weekend of "getting my head into something yet unknown".